pandas/_libs/tslib.pyx

```diff
@@ -142,16 +149,19 @@ def ints_to_pydatetime(ndarray[int64_t] arr, tz=None, freq=None, box=False):
         ndarray[object] result = np.empty(n, dtype=object)
         object (*func_create)(int64_t, pandas_datetimestruct, object, object)
 
-    if box and is_string_object(freq):
-        from pandas.tseries.frequencies import to_offset
-        freq = to_offset(freq)
-
-    if box:
-        func_create = create_timestamp_from_ts
+    if kind == "date":
+        func_create = create_date_from_ts
     else:
-        func_create = create_datetime_from_ts
+        if box and is_string_object(freq):
+            from pandas.tseries.frequencies import to_offset
+            freq = to_offset(freq)
 
-    if tz is not None:
```

Comment:
use the box param and have it take strings instead of boolean

iow box can be date/datetime/timestamp 

you can assert that it is one of those 

you will have to change places where this function. is called (eg box=True -> timestamp) etc